### PR TITLE
boot: move sealing code to fdestate backend

### PR DESCRIPTION
**Pull request by valentindavid**

### CODE REVIEW

File: boot/seal.go [hunk 1]
```diff
@@ -556,9 +394,66 @@ func modesForSystems(modeenv *Modeenv) map[string][]string {
 	return systemToModes
 }
 
+func recoveryBootChainsForSystems(systems []string, modesForSystems map[string][]string, trbl bootloader.TrustedAssetsBootloader, modeenv *Modeenv, includeTryModel bool, seedDir string) (chains []BootChain, err error) {
+	if trbl == nil {
+		return recoveryBootChainsForSystemsUntrusted(systems, modesForSystems, modeenv, includeTryModel, seedDir)
+	}
+
+	return recoveryBootChainsForSystemsTrusted(systems, modesForSystems, trbl, modeenv, includeTryModel, seedDir)
```
Comment (pedronis) [inline boot/seal.go:402]: maybe WithTrustedAssets and WithoutTrustedAssets instead of Trusted/Untrusted ?

File: boot/seal.go [hunk 2]
```diff
@@ -556,9 +394,66 @@ func modesForSystems(modeenv *Modeenv) map[string][]string {
...
+		if err := chainsForModel(modeenv.TryModelForSealing()); err != nil {
+			return nil, err
+		}
+	}
+
+	return chains, nil
+}
+
 // TODO:UC20: this needs to take more than one model to accommodate the remodel
 // scenario
```
Comment (pedronis) [inline boot/seal.go:455]: I think this TODO has been actually implemented by the TryModel support?

Comment (pedronis) [reply boot/seal.go:455]: I think there might be another similar TODO in the file that can also be removed

File: overlord/fdestate/backend/seal.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+	if err := device.StampSealedKeys(params.InstallHostWritableDir, device.SealingMethodFDESetupHook); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func sealKeyForBootChainsBackend(method device.SealingMethod, key, saveKey secboot.BootstrappedContainer, params *boot.SealKeyForBootChainsParams) error {
+	if method == device.SealingMethodFDESetupHook {
+		fmt.Printf("BD\n")
```
Comment (pedronis) [inline overlord/fdestate/backend/seal.go:173]: there are some debug prints  here and below

**Review by pedronis — COMMENTED** [review]
looks good overall, did an initial pass, some small comments

File: boot/export_test.go [hunk 4]
```diff
@@ -206,7 +183,7 @@ func SetBootFlagsInBootloader(flags []string, rootDir string) error {
 }
 
 func (b *BootChain) SecbootModelForSealing() secboot.ModelForSealing {
-	return b.modelForSealing()
+	return b.ModelForSealing()
 }
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply boot/export_test.go:187]: I will need to clean this up after splitting the tests. It still needs to be used in 2 packages now.

File: gadget/install/export_secboot_test.go [hunk 5]
```diff
@@ -41,7 +41,7 @@ func MockSecbootFormatEncryptedDevice(f func(key keys.EncryptionKey, encType sec
 }
 
 func MockBootRunFDESetupHook(f func(req *fde.SetupRequest) ([]byte, error)) (restore func()) {
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply gadget/install/export_secboot_test.go:43]: I wonder if it is used anywhere...

**Review by pedronis — APPROVED** [review]
thank you

File: boot/seal.go [hunk 6]
```diff
@@ -134,91 +123,74 @@ func sealKeyToModeenvImpl(key, saveKey secboot.BootstrappedContainer, model *ass
...
 	}
-
-	if flags.StateUnlocker != nil {
-		relock := flags.StateUnlocker()
-		defer relock()
+	if err := sealKeyToModeenvForMethod(method, key, saveKey, model, modeenv, flags); err != nil {
+		return err
 	}
-	return sealKeyToModeenvUsingSecboot(key, saveKey, model, modeenv, flags)
+	return nil
```
Comment (alfonsosanchezbeato) [inline boot/seal.go:138]: nitpick: is there any reason for not doing directly `return sealKeyToModeenvUsingSecboot(...)`?

File: boot/seal.go [hunk 7]
```diff
@@ -669,11 +561,38 @@ func recoveryBootChainsForSystems(systems []string, modesForSystems map[string][
 	return chains, nil
 }
 
-func runModeBootChains(rbl, bl bootloader.Bootloader, modeenv *Modeenv, cmdlines []string, runSnapsDir string) ([]BootChain, error) {
-	tbl, ok := rbl.(bootloader.TrustedAssetsBootloader)
-	if !ok {
-		return nil, fmt.Errorf("recovery bootloader doesn't support trusted assets")
+func runModeBootChains(rbl bootloader.TrustedAssetsBootloader, bl bootloader.Bootloader, modeenv *Modeenv, cmdlines []string, runSnapsDir string) ([]BootChain, error) {
+	if rbl == nil {
+		return runModeBootChainsUntrusted(modeenv, runSnapsDir)
```
Comment (alfonsosanchezbeato) [inline boot/seal.go:566]: Previously we were returning errors when trusted assets were not supported, what does secboot for the untrusted case? Is this for when we use hooks and there is no TPM?

**Review by alfonsosanchezbeato — COMMENTED** [review]
I have a maybe obvious question

File: boot/seal.go (hunk 6 above)
Comment (valentindavid) [reply boot/seal.go]: Fixed.

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks

**Pull request merged**